README.md
$Id$

Qik site framework - installation
---------------------------------

Just copy the contents of the tarball to the place you want them to have.
This doesn't have to be the root of your web server. For example:

tar xjf qik_release.tar.bz2
mv qik_root/* /var/www/

If you install Qik to be your main website, you can let it handle your Apache
error pages too. For example, add the following to your Apache's config file:

ErrorDocument 401 /page/error/401/
ErrorDocument 403 /page/error/403/
ErrorDocument 404 /page/error/404/
ErrorDocument 500 /page/error/500/


More information
----------------
For more information, see the Qik page on
http://aquariusoft.org/page/html/qik/


core/CHANGELOG
Qik site framework: CHANGELOG

2006-03-31
- Abstrahized data-retrieval tasks. Put all file system related calls into
  their own library [mod_qik_files.php]


2006-03-30
- Finished "logging to file" support


2006-03-29
- Added logging support


2006-02-01
- Better redirection support for faulty uri's
- Shouldn't redirect when being used for the webserver's error pages


2006-01-22
- Added support for "icon" tags, with @@@icon=spreadsheet@@@ etc for instant 24x24
  pictograph
- Fixed lack of tagging support in Sitemap
- Fixed issue with empty items showing up in Sitemap for pages with an empty value
  [which won't show up in the normal navigation]


2006-01-21
- Added support for images in the navigation items
- Added support for more characters in section- and page-keys [for uri's], not
  only A-z and 0-9
- Framework can show generic HTTP Error pages now [config your webserver by setting
  lines like:
  ErrorDocument 403 /page/error/403/


2006-01-19
- Updated Qik so it supports multiple languages, news items and more.
  - tags support 'tiles' and 'news' now, in the form of @@@tile=name@@@ and
    @@@news=name@@@
- Extensive rewrite to make it a lot more versatile and even easier to set up
- Placed all user customizable things inside site/
  - including the pagetemplate.php, which has the functions for generating the
    body, navigation and subnavigation of the pages



core/index.php
<?php

/**
 * index.php Main file for the Qik framework
 *
 * Qik is the legal property of its developer, Michiel Scholten
 * [mbscholtNOSPAM@aquariusoft.org]
 * Please refer to the COPYRIGHT file distributed with this source distribution.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */


/*** Initializing ***/

$skel['version'] = '0.1.16 2006-04-03';
$skel['starttime'] = microtime();

//error_reporting( E_ERROR | E_WARNING | E_PARSE | E_NOTICE );	// set all on
error_reporting( E_ALL );
//error_reporting(0);	// set all off

/* Website configuration */
include_once('config.php');
/* Dictionary for the messages used by the framework */
include_once('modules/dictionary.php');

if ($skel['usedb'])
{
	include_once('modules/mod_qik_db.php');
} else
{
	include_once('modules/mod_qik_files.php');
}

/* Module with generic helpful functions */
include_once('modules/mod_toolkit.php');
/* Module for actually building parts of the pages */
include_once('modules/mod_site.php');
/* Module for expanding tags like @@NEWS@@ to the actual news */
include_once('modules/mod_tagsections.php');

/* Module for logging support */
include_once('modules/mod_logging.php');

include_once('site/pagetemplate.php');


$skel['base_uri'] = dirname($_SERVER['PHP_SELF']) . '/';
if ('//' == $skel['base_uri'])
{
	/* Site is located in the root, compensate for the extra slash */
	$skel['base_uri'] = '/';
}
$url_pieces = parse_url(getenv('SCRIPT_URI'));
$skel['base_server'] = '';
if (!isset($url_pieces['scheme']))
{
	//$url_pieces = parse_url($_SERVER['SCRIPT_URI']);
	$skel['base_server'] = 'http://' . $_SERVER['SERVER_NAME'];
} else
{
	$skel['base_server'] = $url_pieces['scheme'] . '://' . $url_pieces['host'];
}

$sections = getSections($skel);
if (null == $sections)
{
	echo 'Can\'t read site description file!';
	exit;
}

/*** Getting base stuff ***/

/* User should provide the site/pagetemplate.php */
$navbar = buildNav($skel, $sections);

$section = getRequestParam('section', null);
$page = getRequestParam('page', null);
$action = getRequestParam('action', null);


/*
 * rewrite http://example.com/?section=w00t&page=blah to
 * http://example.com/page/w00t/blah/
 */
//$url_pieces = parse_url(getenv('REQUEST_URI'));
//if ('/index.php' == substr($url_pieces['path'], 0, 10) && isset($url_pieces['query']) && '' != $url_pieces['query'])
if (isset($url_pieces['query']) && '' != $url_pieces['query'])
{
	/* Redirect. Check for url injections */
	header('HTTP/1.1 301 Moved Permanently');
	//$redir = 'http://' . $skel['hostname'] . $skel['base'] . 'page/' . $section_name . '/';
	//$redir = 'http://aquariusoft.org' . $skel['base_uri'] . 'page/' . $section . '/';
	$redir = $skel['base_server'] . $skel['base_uri'] . 'page/' . $section . '/';
	if ($section == null || $section == '')
	{
		//$redir = 'http://aquariusoft.org/';
		$redir = $skel['base_server'] . '/';
	} else if ($page != NULL)
	{
		$redir .= $page . '/';
	}
	header('Location: ' . $redir);
	addToLog($skel, $section, $page, 301);
	exit;
//} else if (substr($url_pieces['path'], strlen($url_pieces['path']) - 6, 6) == '/page/')
} else if ($url_pieces['path'] == '/page/')
{
	header('HTTP/1.1 301 Moved Permanently');
	//$redir = 'http://aquariusoft.org/';
	$redir = $skel['base_server'] . '/';
	header('Location: ' . $redir);
	addToLog($skel, '', '', 301);
	exit;
}

if ('sitemap' == $action)
{
	$body = '<h1>' . dict($skel, 'sitemap') . "</h1>\n";
	$body .= buildSitemap($skel, $sections);
	addToLog($skel, 'special', 'sitemap', 200);
	echo processTags($skel, buildPage($skel, dict($skel, 'sitemap'), $navbar, null, $body));
	exit;
} else if ('viewlog' == $action)
{
	$logaction = getRequestParam('logaction', 'pages');
	$offset = getRequestParam('offset', 0);
	$body = '<h1>' . dict($skel, 'visitslog') . "</h1>\n";

	$skel['section'] = dict($skel, 'visitslog');
	$subsections = getLogTypes($skel);
	$subnav = buildSubnav($skel, 'viewlog', $subsections);

	$date = getRequestParam('date', null);
	$body .= buildVisitsLogOverview($skel, $logaction, $offset, $date);
	addToLog($skel, 'special', 'viewlog', 200);
	echo processTags($skel, buildPage($skel, dict($skel, 'visitslog'), $navbar, $subnav, $body));
	exit;
}

if ($section == 'error')
{
	if ($page == '404')
	{
		addToLog($skel, $section, $page, 404);
		echo build404($skel, $navbar);
		/*} else if ($page == '401')
		  {
		 */
	} else {
		//echo buildPage($skel, dict($skel, 'unknown_server_error'), $navbar, $subnav, $body);
		addToLog($skel, $section, $page, 500);
		echo buildHTTPErrorPage($skel, $navbar, $page);
	}
	exit;
}


if ($section == null)
{
	$section = getItem($sections, 0);
}

$section_name = getName($sections, $section);
if ($section_name == null)
{
	addToLog($skel, $section, $page, 404);
	echo build404($skel, $navbar, dict($skel, 'section_not_found'));
	exit;
}

$skel['section'] = $section_name;

$subsections = getSubsections($skel, $section);
$subnav = null;
if (null != $subsections)
{
	$subnav = buildSubnav($skel, $section, $subsections);
}

if (null == $page)
{
	/* If no page is specified, get first page in the sub category list */
	$page = getItem($subsections, 0);
}

$content = getFileContents($skel, $section, $page);
if (null == $content)
{
	addToLog($skel, $section, $page, 404);
	echo build404($skel, $navbar, dict($skel, 'page_x_not_found', $page));
	exit;
}

$page_name = '';
if ($page != null)
{
	$page_name = getName($subsections, $page);
} else
{
	$page_name = $section_name;
}

$body = '';
if (trim($page_name) != '')
{
	$body = '<h1>' . $page_name . "</h1>\n";
}
$body .= $content;

/* Rebuild navbar, now with highlighted section */
$navbar = buildNav($skel, $sections);

addToLog($skel, $section, $page, 200);
echo processTags($skel, buildPage($skel, $page_name, $navbar, $subnav, $body));
?>
